lib/Containers/MerkleTree.cpp

```diff
@@ -1038,162 +951,317 @@ template <typename Hasher, std::uint64_t const BranchingBits>
 typename MerkleTree<Hasher, BranchingBits>::Node const&
 MerkleTree<Hasher, BranchingBits>::node(std::uint64_t index) const noexcept {
   // not thread-safe, lock buffer from outside
-  TRI_ASSERT(index < nodeCountUpToDepth(meta().maxDepth));
+  TRI_ASSERT(index < nodeCountAtDepth(meta().depth));
   uint8_t const* ptr = _buffer.get() + MetaSize + (NodeSize * index);
   return *reinterpret_cast<Node const*>(ptr);
 }
 
 template <typename Hasher, std::uint64_t const BranchingBits>
-std::uint64_t MerkleTree<Hasher, BranchingBits>::index(std::uint64_t key,
-                                                       std::uint64_t depth) const noexcept {
+std::uint64_t MerkleTree<Hasher, BranchingBits>::index(std::uint64_t key) const noexcept {
   // not thread-safe, lock buffer from outside
-  TRI_ASSERT(depth <= meta().maxDepth);
   TRI_ASSERT(key >= meta().rangeMin);
   TRI_ASSERT(key < meta().rangeMax);
 
-  // special fast case
-  if (depth == 0) {
-    return 0;
-  }
-
   std::uint64_t offset = key - meta().rangeMin;
   std::uint64_t chunkSizeAtDepth =
       (meta().rangeMax - meta().rangeMin) /
-      (static_cast<std::uint64_t>(1) << (BranchingBits * depth));
+      (static_cast<std::uint64_t>(1) << (BranchingBits * meta().depth));
   std::uint64_t chunk = offset / chunkSizeAtDepth;
 
-  return chunk + nodeCountUpToDepth(depth - 1);
+  return chunk;
 }
 
 template <typename Hasher, std::uint64_t const BranchingBits>
 void MerkleTree<Hasher, BranchingBits>::modify(std::uint64_t key, bool isInsert) {
   // not thread-safe, shared-lock buffer from outside
   Hasher h;
   std::uint64_t const value = h.hash(key);
-  for (std::uint64_t depth = 0; depth <= meta().maxDepth; ++depth) {
-    bool success = modifyLocal(depth, key, value, isInsert);
-    if (ADB_UNLIKELY(!success)) {
-      // roll back the changes we already made, using best effort
-      for (std::uint64_t d = 0; d < depth; ++d) {
-        [[maybe_unused]] bool rolledBack = modifyLocal(d, key, value, !isInsert);
-      }
-#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
-      checkInternalConsistency();
-#endif
-      throw std::invalid_argument("Tried to remove key that is not present.");
-    }
+ 
+  // adjust bucket node
+  bool success = modifyLocal(key, value, isInsert);
+  if (ADB_UNLIKELY(!success)) {
+    throw std::invalid_argument("Tried to remove key that is not present.");
   }
+    
+  // adjust summary node
+  modifyLocal(meta().summary, 1, value, isInsert);
```

Comment:
It is probably unlikely that this fails, but should we at least add an assertion?

Reply:
I have added such assertion.